app/models/maintenance_tasks/csv_collection_builder.rb

```diff
@@ -3,31 +3,30 @@
 require "csv"
 
 module MaintenanceTasks
-  # Module that is included into Task classes by Task.csv_collection for
-  # processing CSV files.
+  # Strategy for building a Task that processes CSV files.
   #
   # @api private
-  module CsvCollection
-    # The contents of a CSV file to be processed by a Task.
-    #
-    # @return [String] the content of the CSV file to process.
-    attr_accessor :csv_content
-
+  class CsvCollectionBuilder
     # Defines the collection to be iterated over, based on the provided CSV.
     #
     # @return [CSV] the CSV object constructed from the specified CSV content,
     #   with headers.
-    def collection
-      CSV.new(csv_content, headers: true)
+    def collection(task)
+      CSV.new(task.csv_content, headers: true)
     end
 
     # The number of rows to be processed. Excludes the header row from the count
     # and assumed a trailing new line in the CSV file. Note that this number is
     # an approximation based on the number of new lines.
     #
     # @return [Integer] the approximate number of rows to process.
-    def count
-      csv_content.count("\n") - 1
+    def count(task)
+      task.csv_content.count("\n") - 1
+    end
+
+    # Return that the Task processes CSV content.
+    def has_csv_content?
```

Comment:
I think we should have another name for this since we could have CSV content in another way (like we will do in Shopify/shopify for example). This hook method would make more sense as `needs_csv_upload?` or something. But I see why you called it that way, since we have a `csv_content` attribute. Maybe it's just the comment that needs to be changed: "Returns that the Task processes the CSV content passed to the Run"

---

I think looking back, what the gem provides and what the conditionals we've been using (`task_class < CsvCollection` and `task.respond_to?(:csv_content=)` mean in a more generic way is:

* a `input[type=file]` in the view
* an attribute on the Task instance that holds the content† of the file that was uploaded for this run

That way we would not even be prescribing CSV, it could be JSON or whatever, as long as the task has a collection strategy providing items to be processed. It would be more involved to fix this though

† ideally it would be a stream, an IO of some sort, that way we wouldn't have to download the whole file for a single job execution, since with a big file there would be multiple job executions on multiple machines, downloading the whole file each time would be a bit wasteful, but we don't have proper streaming in Active Storage (there is [Service#download_chunk,](https://api.rubyonrails.org/classes/ActiveStorage/Service.html#method-i-download_chunk) but it doesn't give an IO, [Blob#open](https://api.rubyonrails.org/classes/ActiveStorage/Blob.html#method-i-open) does but it downloads everything.